### PR TITLE
[FLINK-1398] Introduce extractSingleField() in DataSet

### PR DESCRIPTION
**Pull request by FelixNeutatz**
This is a prototype how we could implement extractSingleField() for DataSet. Let's discuss :)

You would apply the function like this:

DataSet<?> dataT = env.fromElements(new Tuple2<Integer, Double>(1, 2.0));

dataT.extractSingleField(1, Double.class).print(); //prints 2.0

### CODE REVIEW

**Comment by zentol** [comment]
Why did you make a new operator instead of implementing it as a simple map function?

**Comment by FelixNeutatz** [comment]
Good question, I thought it would be nice to have it implemented "natively", but I can also setup the same thing using a simple map function if this is preferred

**Comment by FelixNeutatz** [comment]
The simple map function implementation reduces the code to 26 lines. I think that this is more reasonable :)

**Comment by StephanEwen** [comment]
I like the idea of having good utilities.

There is always a bit of a trade-off between having the right utilities, and adding so many functions that the API becomes to cluttered and hard to use.

There is also the possibility to have something like a "lib" package of common functions, that one can then simply use. Not quite as convenient as having a function on the `DataSet`, but also prevents overloading the core set of data set operations too much.

**Comment by FelixNeutatz** [comment]
I am really open about the place of this functionality, but I'd really love to have it. Shall we start a package in flink-java?

**Comment by rmetzger** [comment]
+1 for some utilities.
I'm not sure however where to put it.
Should we add another maven module? Make it part of the current "flink-java" ? Or start it as a github repo outside of the main project?

**Comment by StephanEwen** [comment]
This looks like a nice approach.

Concerning the naming/packaging: How about we create a package `org.apache.flink.api.java.lib` with various classes with utility functions? Such as
  - _Accessors_: Get fields from tuples / pojos / ...
  - _Tuple Utils_: Wrap / unwrap as Tuple1, duplicate values, reverse, whatever...
  - ...

**Comment by zentol** [comment]
@StephanEwen would that look like this?
```java
reverse(dataset.map(...)).filter(...)
```

Ive been thinking about the API overload issue a bit: could a DataSet have a field that offered more functions? something like:
```java
dataset.map().ext.reverse().filter()
```

File: flink-java/src/main/java/org/apache/flink/api/java/lib/DataSetUtil.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		TupleTypeInfo<IN> tupleInfo = (TupleTypeInfo) ds.getType();
+		if(fieldIndex >= tupleInfo.getArity() || fieldIndex < 0) {
+			throw new IndexOutOfBoundsException("The field index has to be between 0 and " + (tupleInfo.getArity()-1));
+		}
+		
+		if(!tupleInfo.getTypeAt(fieldIndex).equals(TypeExtractor.createTypeInfo(outputType))) {
+			throw new IllegalArgumentException("The output class type has to be: " + tupleInfo.getTypeAt(fieldIndex).toString());
+		}
+		
+		return ds.map(new ExtractElement(fieldIndex)).returns(tupleInfo.getTypeAt(fieldIndex));
```
Comment (fhueske) [inline flink-java/src/main/java/org/apache/flink/api/java/lib/DataSetUtil.java]: Add `.name("Extract Field "+fieldIndex)` to specify the name of the Map operator.

**Comment by fhueske** [comment]
How do we proceed with this PR?
I think it looks good and would be OK with adding this feature to a `DataSetUtils` class.

Other opinions?

**Comment by StephanEwen** [comment]
I think this is good in general, modulo two issues:
  - There are going to be more Utils, to I would like to give it a more speaking name, like "TupleUtils", or "TupleExtractors", or something along these lines.
  - I think we can omit the return type class. Similar as with the projections, this should not be needed.

**Comment by aljoscha** [comment]
I'm against merging this, to me this seems like a lot of code for something that can be achieved using a simple mapper. Also, if we add this, then we should also have a version for streaming, so that would mean adding a DataStreamUtils, or have TupleUtils serve both DataSet and DataStream.

In general I'm against all the specialised code for dealing with Tuples, such as the aggregation code and the join/projection code.

But maybe that's just me... What do you think of my comments?

**Comment by rmetzger** [comment]
I would put this into the `flink-contrib` module.

**Comment by fhueske** [comment]
@aljoscha Those are good points. I like the idea of cleaning the Java and Scala APIs from operations that work on structured data such as projection, aggregation, etc. and support those use cases through the Table API. Not sure if the Table API can serve as a complete replacement at this point in time, but moving it there is the right thing to do, IMO.

But this discussion should happen on the dev mailing list, not in some PR comment thread ;-)

**Comment by aljoscha** [comment]
Yes, I think we should start a discussion there. I just wanted to give the reasons for my opinion here.

**Comment by StephanEwen** [comment]
+1 for not specializing code in the `DataSet`

A `TupleUtil` would be fine, in my opinion

**Comment by FelixNeutatz** [comment]
So, what is the final decision here?

>  this seems like a lot of code for something that can be achieved using a simple mapper
--> that is exactly the reason why I would want this functionality - it is too much code for a simple thing :)

**Comment by aljoscha** [comment]
I think the consensus was that we don't want to have such a method in the DataSet API. We can, however, put a utility for this in flink-contrib. This utility should work for both batch and streaming? Any other opinions? Please correct me if I'm wrong.

**Comment by StephanEwen** [comment]
I think the options were a to have a `TupleUtil` class either in 
  - `flink-contirib`
  - `flink-java` in a `util` or `lib` subfolder.

I would vote for contrib for now.

**Comment by fhueske** [comment]
+1 for contrib

**Comment by tonycox** [comment]
Hi @FelixNeutatz Can you finish this PR ?

**Comment by StephanEwen** [comment]
Playing devil's advocate here: Given that this is really easy to just solve on-the-fly in a program (especially with Java 8 and lambdas) has not been requested another time by the community, I would suggest to actually not add this to Flink.

I think that in this case, the benefit of reducing maintenance effort outweighs the benefit to users via this utility.

**Comment by fhueske** [comment]
I agree with Stephan

**Comment by tonycox** [comment]
Okay, I'll add it to clean up jira list

**Pull request closed without merging**